File: VB/T401286/Controllers/HomeController.vb
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Web
Imports System.Web.Mvc
Imports Models
Imports DevExpress.Web.Mvc

Namespace T401286.Controllers
    Public Class HomeController
        Inherits Controller

        Public Function Index() As ActionResult
            ' DXCOMMENT: Pass a data model for GridView

            Return View()
        End Function

        Public Function GridViewPartialView() As ActionResult
            ' DXCOMMENT: Pass a data model for GridView in the PartialView method's second parameter
            Return PartialView("GridViewPartialView", Company.OfficeRooms)
        End Function

        <HttpPost, ValidateInput(False)> _
        Public Function BatchUpdateRoomsPartial(ByVal batchValues As MVCxGridViewBatchUpdateValues(Of Room, Integer)) As ActionResult
            For Each item In batchValues.Insert
                If batchValues.IsValid(item) Then
                    Company.InsertRoom(item)
                Else
                    batchValues.SetErrorText(item, "Correct validation errors")
                End If
            Next item
            For Each item In batchValues.Update
                If batchValues.IsValid(item) Then
                    Company.UpdateRoom(item)
                Else
                    batchValues.SetErrorText(item, "Correct validation errors")
                End If
            Next item
            For Each itemKey In batchValues.DeleteKeys
                Company.RemoveRoomByID(itemKey)
            Next itemKey
            Return PartialView("GridViewPartialView", Company.OfficeRooms)
        End Function

    End Class
End Namespace

File: CS/T401286/Controllers/HomeController.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Web;
using System.Web.Mvc;
using Models;
using DevExpress.Web.Mvc;

namespace T401286.Controllers
{
    public class HomeController : Controller
    {
        public ActionResult Index()
        {
            // DXCOMMENT: Pass a data model for GridView

            return View();    
        }
        
        public ActionResult GridViewPartialView() 
        {
            // DXCOMMENT: Pass a data model for GridView in the PartialView method's second parameter
            return PartialView("GridViewPartialView", Company.OfficeRooms);
        }

        [HttpPost, ValidateInput(false)]
        public ActionResult BatchUpdateRoomsPartial(MVCxGridViewBatchUpdateValues<Room, int> batchValues)
        {
            foreach (var item in batchValues.Insert)
            {
                if (batchValues.IsValid(item))
                    Company.InsertRoom(item);
                else
                    batchValues.SetErrorText(item, "Correct validation errors");
            }
            foreach (var item in batchValues.Update)
            {
                if (batchValues.IsValid(item))
                    Company.UpdateRoom(item);
                else
                    batchValues.SetErrorText(item, "Correct validation errors");
            }
            foreach (var itemKey in batchValues.DeleteKeys)
            {
                Company.RemoveRoomByID(itemKey);
            }
            return PartialView("GridViewPartialView", Company.OfficeRooms);
        }
    
    }
}